rp2040-hal/src/pio.rs

```diff
@@ -233,72 +371,180 @@ impl<P: Instance> StateMachine<P> {
+        let is_empty =
+            unsafe { &*self.sm.block }.fstat.read().rxempty().bits() & (1 << self.sm.id) != 0;
 
         if is_empty {
             return None;
         }
 
-        Some(self.block().rxf[self.id as usize].read().bits())
+        // Safety: The register is unique to this state machine.
+        Some(
+            unsafe { &*self.sm.block }.rxf[self.sm.id as usize]
+                .read()
+                .bits(),
+        )
     }
 
     /// Write an element to TX FIFO.
     ///
     /// Returns `true` if the value was written to FIFO, `false` otherwise.
-    pub fn write_tx(&self, value: u32) -> bool {
-        let is_full = self.block().fstat.read().txfull().bits() & (1 << self.id) != 0;
+    pub fn write_tx(&mut self, value: u32) -> bool {
+        // Safety: The register is never written by software.
+        let is_full =
+            unsafe { &*self.sm.block }.fstat.read().txfull().bits() & (1 << self.sm.id) != 0;
 
         if is_full {
             return false;
         }
 
-        self.block().txf[self.id as usize].write(|w| unsafe { w.bits(value) });
+        // Safety: The register is unique to this state machine.
+        unsafe { &*self.sm.block }.txf[self.sm.id as usize].write(|w| unsafe { w.bits(value) });
 
         true
     }
 }
 
+impl<P: PIOExt> StateMachine<P, Stopped> {
+    /// Starts execution of the selected program.
+    pub fn start(mut self) -> StateMachine<P, Running> {
+        // Enable SM
+        self.sm.set_enabled(true);
+
+        StateMachine {
+            sm: self.sm,
+            program: self.program,
+            _phantom: core::marker::PhantomData,
+        }
+    }
+
+    /// Sets the pin directions for the specified pins.
+    ///
+    /// The `pins` parameter specifies a set of pins as a mask, and `pindir` contains the
+    /// directions that are configured for these pins. The bits in both masks correspond to the pin
+    /// number. The user has to make sure that they do not select any pins that are in use by any
+    /// other state machines of the same PIO block.
+    ///
+    /// This function needs to be called for sideset pins if they are supposed to be used as
+    /// output pins.
+    pub fn set_pindirs_with_mask(&mut self, mut pins: u32, pindir: u32) {
+        let mut pin = 0;
+        let prev_pinctrl = self.sm.sm().sm_pinctrl.read().bits();
+        // For each pin in the mask, we select the pin as a SET pin and then execute "set PINDIRS,
+        // <direction>".
+        while pins != 0 {
+            if (pins & 1) != 0 {
+                self.sm.sm().sm_pinctrl.write(|w| {
+                    unsafe {
+                        w.set_count().bits(1);
+                        w.set_base().bits(pin as u8);
+                    }
+                    w
+                });
+                self.sm.sm().sm_instr.write(|w| {
+                    unsafe {
+                        const SET_PINDIRS: u16 = 0xe080;
```

Comment:
We should be able to use instruction encoder here. See https://github.com/rp-rs/rp-hal/blob/833b69819e0d91b708c1890c25d0f7ecab0964a1/rp2040-hal/src/pio.rs#L168

Reply:
The code got slightly longer - is this what you mean?

Reply:
Yes, this is what I meant.